File: src/com/leetcode/p707/Solution.java
package com.leetcode.p707;

class Node {
    public int val;
    public Node next;
    public Node(int val) {
        this.val = val;
    }

}
/**
 * 单链表
 *
 * 单纯的链表处理是比较简单的，麻烦的是遇到几个特殊的节点时的处理：
 *
 * - 添加到头节点，删除头节点，注意 header 指针的处理
 * - 删除到尾结点，需要将 tail 指针指向前一个节点
 */
class MyLinkedList {
    private Node header;
    private Node tail;
    int size = 0;
    
    public MyLinkedList() {
        header = null;
        tail = null;
    }

    public int get(int index) {
        if (index >= size || index < 0) {
            return -1;
        }

        int i = 0;
        Node node = header;
        while (i != index) {
            node = node.next;
            i ++;
        }
        return node.val;
    }

    public void addAtHead(int val) {
        Node newHead= new Node(val);
        newHead.next = header;
        header = newHead;
        size ++;
        if (tail == null) {
            tail = newHead;
        }
    }

    public void addAtTail(int val) {
        if (header == null) {
            addAtHead(val);
            return;
        }

        Node newTail = new Node(val);
        tail.next = newTail;
        tail = newTail;
        size++;
    }

    public void addAtIndex(int index, int val) {
        if (index < 0 || index > size) {
            return;
        }

        if (index == 0) {
            addAtHead(val);
            return;
        }

        if(index == size) {
            addAtTail(val);
            return;
        }

        int i = 1;
        Node previous = header;
        Node node = header.next;
        while (i != index) {
            node = node.next;
            previous = previous.next;
            i ++;
        }

        Node newNode = new Node(val);
        newNode.next = node;
        previous.next = newNode;
        size ++;
    }

    public void deleteAtIndex(int index) {
        if (index < 0 || index >= size) {
            return;
        }

        if (index == 0) {
            header = header.next;
            if (size == 1) {
                tail = header;
            }
            size --;
            return;
        }

        int i = 1;
        Node node = header.next;
        Node previous = header;
        while (i != index) {
            node = node.next;
            previous = previous.next;
            i ++;
        }

        if (index == size - 1) {
            previous.next = null;
            tail = previous;
        }else {
            previous.next = node.next;
        }


        size --;
    }

}

File: src/com/leetcode/p707/solution.go
package main

type Node struct {
	Val  int
	Next *Node
	Pre  *Node
}

type MyLinkedList struct {
	Size   int
	header *Node
	tail   *Node
}

/*
*
双向链表实现

1. 每次节点变化需要变化 Pre、Next 两个指针
2. edge case: 新增 header、tail; 删除 header、tail; heaer 和 tail 是同一个
*/
func Constructor() MyLinkedList {
	return MyLinkedList{0, nil, nil}
}

func (this *MyLinkedList) Get(index int) int {
	if index < 0 || index >= this.Size {
		return -1
	}

	node := this.header
	for i := 0; i != index; i++ {
		node = node.Next
	}
	return node.Val
}

func (this *MyLinkedList) AddAtHead(val int) {
	if this.header == nil {
		this.header = &Node{val, nil, nil}
		this.tail = this.header
	} else {
		newHeader := &Node{val, this.header, nil}
		this.header.Pre = newHeader
		this.header = newHeader
	}
	this.Size++

}

func (this *MyLinkedList) AddAtTail(val int) {
	if this.tail == nil {
		this.AddAtHead(val)
		return
	}

	newTail := &Node{val, nil, this.tail}
	this.tail.Next = newTail
	this.tail = newTail
	this.Size++
}

func (this *MyLinkedList) AddAtIndex(index int, val int) {
	if index < 0 || index > this.Size {
		return
	}
	if index == 0 {
		this.AddAtHead(val)
		return
	}

	if index == this.Size {
		this.AddAtTail(val)
		return
	}

	node := this.header.Next
	for i := 1; i != index; i++ {
		node = node.Next
	}

	newNode := &Node{val, node, node.Pre}
	node.Pre.Next = newNode
	node.Pre = newNode
	this.Size++
}

func (this *MyLinkedList) DeleteAtIndex(index int) {
	if index < 0 || index >= this.Size {
		return
	}
	if index == 0 {
		this.deleteHeader()
		return
	}
	if index == this.Size-1 {
		this.deleteTail()
		return
	}

	node := this.header.Next
	for i := 1; i != index; i++ {
		node = node.Next
	}

	node.Next.Pre = node.Pre
	node.Pre.Next = node.Next
	node.Next = nil
	node.Pre = nil
	this.Size--

}

func (this *MyLinkedList) deleteHeader() {
	if this.tail == this.header {
		this.header = nil
		this.tail = nil
	} else {
		oldHeader := this.header
		this.header.Next.Pre = nil
		this.header = this.header.Next
		oldHeader.Next = nil
	}
	this.Size--
}

func (this *MyLinkedList) deleteTail() {
	if this.tail == this.header {
		this.header = nil
		this.tail = nil
	} else {
		oldTail := this.tail
		this.tail.Pre.Next = nil
		this.tail = this.tail.Pre
		oldTail.Pre = nil
	}
	this.Size--
}

func main() {
	obj := Constructor()
	obj.AddAtHead(1)
	obj.AddAtTail(3)
	obj.AddAtIndex(3, 2)
	obj.Get(1)
	obj.DeleteAtIndex(1)
	obj.Get(1)

}
